axiom/tests/test_log_events.c

```diff
@@ -0,0 +1,169 @@
+#define LOG_TIMESTAMP 12345
+
+static nr_log_event_t* create_sample_event(const char* message) {
+  nr_log_event_t* e = nr_log_event_create();
+  nr_log_event_set_log_level(e, LOG_LEVEL);
+  nr_log_event_set_message(e, message);
+  nr_log_event_set_timestamp(e, LOG_TIMESTAMP * NR_TIME_DIVISOR_MS);
+  return e;
+}
+
+static void test_events_success(void) {
+  /* Normal operation */
+  nr_analytics_events_t* events = NULL;
+  nr_log_event_t* e = NULL;
+  nr_random_t* rnd = nr_random_create_from_seed(12345);
+  const char* json;
+  const char* expected;
+
+  events = nr_log_events_create(10);
+  tlib_fail_if_null("events created", events);
+
+  e = create_sample_event(LOG_MESSAGE_0);
+  nr_log_events_add_event(events, e, rnd);
+  nr_log_event_destroy(&e);
+
+  e = create_sample_event(LOG_MESSAGE_1);
+  nr_log_events_add_event(events, e, rnd);
+  nr_log_event_destroy(&e);
+
+  e = create_sample_event(LOG_MESSAGE_2);
+  nr_log_events_add_event(events, e, rnd);
+  nr_log_event_destroy(&e);
+
+  e = create_sample_event(LOG_MESSAGE_3);
+  nr_log_events_add_event(events, e, rnd);
+  nr_log_event_destroy(&e);
+
+  tlib_pass_if_int_equal("events number seen updated", 4,
+                         nr_log_events_number_seen(events));
+  tlib_pass_if_int_equal("events number saved updated", 4,
+                         nr_log_events_number_saved(events));
+
+#define TEST_EVENTS_GET(MESSAGE, IDX)                    \
+  do {                                                   \
+    json = nr_log_events_get_event_json(events, IDX);    \
+    tlib_fail_if_null("no json", json);                  \
+    expected = "[{" \
+                 "\"message\":\"" MESSAGE "\"," \
+                 "\"log.level\":\"" LOG_LEVEL "\"," \
+                 "\"timestamp\":" NR_STR2(LOG_TIMESTAMP) "," \
+                 "\"trace.id\":\"null\"," \
+                 "\"span.id\":\"null\"," \
+                 "\"entity.guid\":\"null\"," \
+                 "\"entity.name\":\"null\"," \
+                 "\"hostname\":\"null\"" \
+              "}]";                                    \
+    tlib_pass_if_str_equal("add event", expected, json); \
+  } while (0)
+
+  /* Test events are stored in order */
+  TEST_EVENTS_GET(LOG_MESSAGE_0, 0);
+  TEST_EVENTS_GET(LOG_MESSAGE_3, 3);
+
+  /* Getting events should not remove them from the pool */
+  tlib_pass_if_int_equal("events number saved preserved", 4,
+                         nr_log_events_number_saved(events));
+
+  /* Getting event out of bounds should not crash */
+  json = nr_log_events_get_event_json(events, -1);
+  tlib_pass_if_null("event from out of bounds (lower)", json);
+
+  json = nr_log_events_get_event_json(events,
+                                      nr_log_events_number_saved(events));
+  tlib_pass_if_null("event from out of bounds (higher)", json);
+
+  nr_random_destroy(&rnd);
+  nr_log_events_destroy(&events);
+  tlib_pass_if_null("events destroyed", events);
+}
+
+static void test_events_sample(void) {
```

Comment:
Maybe exercise the ```nr_analytics_events_is_sampling()``` function as well as the return value from ```nr_log_events_add_event()``` indicating sampling occurred?

Reply:
Addressed in https://github.com/newrelic/newrelic-php-agent/commit/d5ff416e2f497e9e5651cd3b5d2667914e422f20.